### PR TITLE
External links refactoring

### PR DESCRIPTION
**Pull request by banukutlu**
Part of #759 

This is to move all external links to the top section to display below the Access Online links as mentioned [here](https://github.com/psu-libraries/psulib_blacklight/issues/759#issuecomment-866189860), also includes a bit of refactoring to change the external links partials to components. This is more of a first pass since these components can be more cleaned up.

There is more todo in terms of the display of the links aka collapsing if there are more than 3 or 5 links. I wanted to get some feedback before adding more changes.

### CODE REVIEW

**Review by awead — CHANGES_REQUESTED** [review]
I really like where this is headed. I've added some suggestions inline in the diff, but that last thing I want to suggest is that I think there's some room for reducing the number of link component classes. I think you could get away with three or less. Define a base class:

``` ruby
module ExternalLinks
  class BaseLinksComponent < ViewComponent::Base
    attr_reader :heading

    def initialize(links:, heading:, show_links: true)
      @links = links
      @heading = heading
      @show_links = show_links
    end
    
    def render?
      links.present? && show_links?
    end

    private
      attr_reader :links
      
      def show_links?
        !!@show_links
      end
  end
end
``` 

Then the view is:

``` ruby
<div class="bs-callout bs-callout-primary">
  <h5><%= heading %></h5>

  <ul class="external-links-access-online">
    <% links.map do |item| %>
      <%= render ExternalLinks::LinkComponent.new(item) %>
    <% end %>
  </ul>
</div>
```

That should cover all of them except Hathi and PsuDigitalCollections. For those, define subclasses for your base:

``` ruby

class HathiLinkComponent < BaseLinksComponent
  def initialize(links:, show_component:)
    @links = links
    @show_component = @show_component
  end
end

class PsuDigitalCollectionsLinksComponent < BaseLinksComponent
end
```

And then have separate html view only for those those two.

It might be overkill... 🤷 But something to play around with.

Another idea is to add the image_tag as an option to the base class and expand the LinkComponent to include the additional hathi keys, then you could have only _one_ links component class for everyting, but that seems a little forced.

**Comment by awead** [comment]
GREAT MINDZ!!!! 🧠 <=> 🧠

File: app/components/external_links/psu_digital_collections_links_component.html.erb [hunk 1]
```diff
@@ -6,8 +6,12 @@
...
 
   <ul class="external-links-psu-digital-collections">
-    <% psu_digital_collections_links.map do |item| %>
-      <li><i class="fas fa-external-link-alt" aria-hidden="true"></i> <%= item['prefix'] %><a href="<%= item['url'] %>"><%= item['text'] %></a><%= item['notes'] %></li>
+    <% links.map do |item| %>
+        <li>
+          <i class="fas fa-external-link-alt" aria-hidden="true"></i> <%= item['prefix'] %>
+          <%= link_to item['text'], item['url'] %>
+          <%= item['notes'] %>
+        </li>
```
Comment (awead) [inline app/components/external_links/psu_digital_collections_links_component.html.erb:14]: Use the ExternalLinks::LinkComponent here.

**Review by awead — CHANGES_REQUESTED** [review]
Fantastic! That really came together nicely. You missed one thing in the collections links component. Otherwise, this is SOLID.

File: app/components/external_links/psu_digital_collections_links_component.html.erb (hunk 1 above)
Comment (banukutlu) [reply app/components/external_links/psu_digital_collections_links_component.html.erb:14]: ah yes, thanks!

**Comment by nicktiberi** [comment]
This looks great! Love seeing our usage of components go up. 🚀 

Not exactly sure what went wrong in the prev/next spec, but once that is fixed, this is good to go from my perspective.

**Comment by banukutlu** [comment]
> This looks great! Love seeing our usage of components go up. 🚀
> 
> Not exactly sure what went wrong in the prev/next spec, but once that is fixed, this is good to go from my perspective.

Thanks Nick, glad that you like the refactoring here. I will also need to add collapsing for when there are 3+ links for each group.

**Comment by banukutlu** [comment]
This is ready for review again, except a feature spec to test the toggle to view more/less links. I cannot find an example record that has more than 3 links for a type. I added a component test though but would be nice to test the toggling in a feature spec. I will check with Ruth and Maryam to see if we can find a record and get that to included as part of the sample.

**Comment by nicktiberi** [comment]
Here's a record that has 4 external links: https://catalog.libraries.psu.edu/catalog/24798226

File: app/javascript/psulib_blacklight/styles/_external_links.scss [hunk 2]
```diff
@@ -8,3 +8,13 @@
...
+
+// scss-lint:disable QualifyingElement
+button.btn.toggle-external-links.collapsed::before {
+  content: 'View More';
+}
+
+button.btn.toggle-external-links::before {
+  content: 'View Less';
+}
+// scss-lint:enable QualifyingElement
```
Comment (nicktiberi) [inline app/javascript/psulib_blacklight/styles/_external_links.scss:20]: You can actually remove the `button` prefix on these selectors — it's not actually needed, and then you can also remove the `scss-lint` comments. `.btn` is enough to target these elements.

Your selectors can just be `.btn.toggle-external-links.collapsed::before` and `.btn.toggle-external-links::before`

**Review by nicktiberi — COMMENTED** [review]
I just have a minor comment about SCSS linting, but overall this looks and works great! I also posted a link to a record with 4 links, so maybe we can pull it in and use it in a feature spec.

File: app/javascript/psulib_blacklight/styles/_external_links.scss (hunk 2 above)
Comment (banukutlu) [reply app/javascript/psulib_blacklight/styles/_external_links.scss:20]: ahh thanks! will do this change

**Comment by banukutlu** [comment]
> Here's a record that has 4 external links: https://catalog.libraries.psu.edu/catalog/24798226

yeasssss! awesome! you rock thanks 🎉  👏🏻  Let me see if I can get this one added to to sample file.

**Review by nicktiberi — APPROVED** [review]

**Pull request merged**